### DISCUSSION TITLE
CH1116 driver

### DISCUSSION THREAD

**Discussion by olikraus** [olikraus/u8g2#2031] [Q&A]
There is a discussion for on the CH1116 driver IC here:

https://github.com/olikraus/u8g2/issues/2030

There might be a constructor for this driver in the future.

Meanwhile use SSD1309 or SH1106 constructor
- CH1116 with external VPP power: Use SSD1309 constructor
- CH1116 with internal VPP generation (Charge Pump): Use SH1106 constructor

Often it is not visible from outside, which constructor is required. If the display only has a flex cable connections, then probably the SH1106 constructor should be used. If the display is mounted on a PCB and the PCB includes a step up converter circuit, then the SSD1309 must be used.

**Comment by JAke596-ja** [comment]
How can I modify the SH1106 code to make it compatible with the CH1116?

**Reply by olikraus** [reply]
From my perspective it should be compatible already. What else should be changed?

**Comment by JAke596-ja** [comment]
I can light up the SSD1306 using the initialization code for the SH1106 with a 128x64 resolution, but I can't get the CH1106 to light up with the same SH1106 configuration code. In theory, the SH1106 configuration should be compatible with the CH1106, but it doesn't seem to be working. I'm not sure what the issue is. Could you help me understand why this is happening?

**Comment by olikraus** [comment]
> Could you help me understand why this is happening?

Probably not. The only person who could tell you what to do is the display vendor. Only the creator of the display can tell you which init code to use.

**Comment by JAke596-ja** [comment]
![c82c436e427ad71064efce8768d5bafc](https://github.com/user-attachments/assets/0a1a2813-2e70-4823-b7da-c346e230ddac)
According to my research, the initialization code for SH1106 only appears in the image mentioned above, and there is no initialization code for OLED anywhere else. How can I adapt it to CH11116 by modifying the initialization commands? Or which part of the code in the u8x8_d_ssd1306_128x64_noname.c file should I modify? Could you please help me with this?

**Comment by olikraus** [comment]
each constructor has its own init code, so for example the ssd1306 128x64 noname init code is here: https://github.com/olikraus/u8g2/blob/2b75f932b5ef4b4de8edf73e1a690702a35b1976/csrc/u8x8_d_ssd1306_128x64_noname.c#L41-L75

You do not need to modify the code itself, you can always update the init code by using u8g2.sendF() command: 
https://github.com/olikraus/u8g2/wiki/u8g2reference#sendf

So for example in the init code the charge pump setup is "U8X8_CA(0x08d, 0x014)", with sendF it looks like this: 
``` C++
u8g2.sendF("ca", 0x8d, 0x14);
```

**Comment by JAke596-ja** [comment]
void u8g2Init(u8g2_t *u8g2)
{
	u8g2_Setup_sh1106_i2c_128x64_noname_f(u8g2, U8G2_R2, u8x8_byte_sw_i2c, u8x8_gpio_and_delay);  
	u8g2_InitDisplay(u8g2); 
	u8g2_SetPowerSave(u8g2, 0); 
	u8g2_ClearBuffer(u8g2);
}
int main(void)
{	    		    	
	IIC_Init();
	u8g2Init(&u8g2);
	u8g2_SetFontMode(&u8g2, 1);
  u8g2_SetFont(&u8g2, u8g2_font_unifont_t_symbols);
while(1)
{ 
}
}
As for the code mentioned above, if I select the SH1106 constructor in the main function and choose the SH1106 I2C constructor u8g2_Setup_sh1106_i2c_128x64_noname_f in u8g2_setup.c, do I need to modify the code in u8x8_d_ssd1306_128x64_noname.c to light up the CH1116 OLED screen, or can it be directly adapted to work with the CH1116 OLED display?Can you help me answer this question?

**Reply by olikraus** [reply]
you need to find the corresponding code, which should be this in your case:
https://github.com/olikraus/u8g2/blob/2b75f932b5ef4b4de8edf73e1a690702a35b1976/csrc/u8x8_d_ssd1306_128x64_noname.c#L479-L499
In line 491 the corresponding init sequence is used, which is here:
https://github.com/olikraus/u8g2/blob/2b75f932b5ef4b4de8edf73e1a690702a35b1976/csrc/u8x8_d_ssd1306_128x64_noname.c#L42
So indeed you could modify the u8x8_d_ssd1306_128x64_noname_init_seq array.
However you could also add your own init commands after the u8g2.begin() by using u8g2.sendF(). I think this would be more reliable and you code will still work even after u8g2 lib updates.
